presto-main/src/main/java/com/facebook/presto/operator/window/WindowPartition.java

```diff
@@ -194,19 +216,62 @@ else if (frameInfo.getType() == RANGE) {
         return new Range(frameStart, frameEnd);
     }
 
-    private boolean emptyFrame(FrameInfo frameInfo, int rowPosition, int endPosition)
+    private int precedingEndRange(long endValue)
+    {
+        int peerGroupEndIndex = peerGroupEndIndices.indexOf(peerGroupEnd);
+        if (peerGroupEndIndex < endValue) {
+            return peerGroupEnd - 1;
+        }
+        return peerGroupEndIndices.get(toIntExact(peerGroupEndIndex - endValue)) - 1;
+    }
+
+    private int precedingStartRange(long startValue)
+    {
+        int peerGroupStartIndex = peerGroupStartIndices.indexOf(peerGroupStart);
+        if (peerGroupStartIndex < startValue) {
+            return 0;
+        }
+        return peerGroupStartIndices.get(toIntExact(peerGroupStartIndex - startValue));
+    }
+
+    private int followingRange(int followingPeerGroupStart, int endPosition, long value)
+    {
+        if (value == 0) {
+            return followingPeerGroupStart;
+        }
+        // TODO: Optimize this to *not* look for peers often, probably have pageIndex keep the peer groups
+        int followingPeerGroupEnd = 0;
+        int currentValue = 0;
+        while (currentValue < value) {
+            boolean peerFound = false;
+            followingPeerGroupEnd = followingPeerGroupStart + 1;
+            while ((followingPeerGroupEnd < partitionEnd) && pagesIndex.positionEqualsPosition(peerGroupHashStrategy, followingPeerGroupStart, followingPeerGroupEnd)) {
```

Comment:
What if there are no peers, just each row is a distinct value within the partition?
You'd never set `peerFound = true`, IIUC so this loop would never end?
Can you explain this to me?